### DISCUSSION TITLE
How to access providers from a flutter app's `main()` function?

### DISCUSSION THREAD

**Discussion by fabis94** [rrousselGit/riverpod#1387] [Q&A] [answered]
It appears from the docs that you can retrieve `ref`s from within a `ProviderScope` builder, but there's no mention of whether it's possible to initialize a ProviderScope earlier in `main()` and get `ref`s out from it without starting to build any widgets. Is it possible?

My use case is that I want to load data from secure storage in `main()` before I start building any widgets, but to do that I have to inject a couple of services which I manage with providers.

**Comment by GitGud31** [comment]
Maybe [this](https://github.com/GitGud31/theme_riverpod_hive/blob/master/lib/app.dart) helps? Load theme from the database before building any widgets.

**Comment by ghost** [comment]
You can use `ProviderContainer` before the main function. And then read the provider from the container. After that wrap the app with `UncontrolledProviderScope` passing the container.

**Reply by ghost** [reply]
```
 
 ​  ​final​ container ​=​ ​ProviderContainer​( 
 ​    observers​:​ [​Logger​()], 
 ​    overrides​:​ [ 
 ​      isarProvider.​overrideWithValue​(isar), 
 ​      secureStorageProvider.​overrideWithValue​(storage) 
 ​    ], 
 ​  ); 
  
 ​  ​// get tokens from secure storage 
 ​  ​await​ container.​read​(tokenNotifierProvider.notifier).​initialize​(); 
  
  
 ​  ​handleError​( 
 ​    () ​=>​ ​runApp​( 
 ​      ​UncontrolledProviderScope​( 
 ​        container​:​ container,

...
```

Example code.

**Reply by fabis94** [reply]
Exactly what I was looking for, thanks!

**Reply by degits** [reply]
As I think, Making Global ProviderContainer is an anti-pattern.

**Reply by mustafa-fahimi** [reply]
Is this still relevent? Just asking because i found this in the docs:
```
//DON'T do this:
final myProviderContainer = ProviderContainer();
void main(){
  runApp(MyApp());
}
```
@rrousselGit

**Reply by kane-knowby** [reply]
Wondering this as well ^

**Comment by rrousselGit** [comment, marked as answer]
Nowadays the recommended approach isn't to put logic in the main. See https://riverpod.dev/docs/essentials/eager_initialization

**Reply by Jacfger** [reply]
I have just come across this exact problem, and I'm kinda confused right now after reading all these materials and a few discussions.

AFAIK, the recommended way to do that before would be what mentioned in comments above, instantiating a `ProviderContainer` ourselves and put it into the `UncontrolledProviderScope` inside the main function. So in case any kind of initialization that we need, we can either do `container.read(...).init` or `container.read(... .future)` to initiate the provider.

But with the EagerInitialization widget, we kinda have to use isLoading state inside the provider to notify the initialization widget that it had loaded properly, doesn't that introduce a lot of boilerplate code especially with a lot of provider with async initialization? What is the actual benefit to do it this way?

@rrousselGit

**Reply by Jacfger** [reply]
Ah my bad, I just realized you've actually answered in this issue.

https://github.com/rrousselGit/riverpod/issues/3518#issuecomment-2106968219

Thanks for your awesome work in any case.

**Reply by vitaliiboiko360** [reply]
The link you shared is no longer working - 
<img width="440" height="394" alt="image" src="https://github.com/user-attachments/assets/72e03abd-104d-47ff-baf9-cc45788499be" />

Maybe it's been moved to  
https://riverpod.dev/docs/how_to/eager_initialization
